src/coins.cpp

```diff
@@ -1063,13 +1247,40 @@ bool CCoinsViewCache::RevertCertOutputs(const CScCertificate& cert, const CTxUnd
     }
 
     CSidechainsMap::iterator scIt = ModifySidechain(scId);
-    scIt->second.scInfo.balance += totalAmount;
-    scIt->second.scInfo.lastEpochReferencedByCertificate = certUndoEntry.replacedLastCertEpoch;
-    scIt->second.scInfo.lastCertificateHash = certUndoEntry.replacedLastCertHash;
-    scIt->second.flag = CSidechainsCacheEntry::Flags::DIRTY;
 
-    LogPrint("cert", "%s():%d - amount restored to scView (amount=%s, resulting bal=%s) %s\n",
-        __func__, __LINE__, FormatMoney(totalAmount), FormatMoney(scIt->second.scInfo.balance), scId.ToString());
+    // restore only if this is the top quality cert for this epoch
+    LogPrint("cert", "%s():%d - cert %s, last cert %s, undo cert %s\n", __func__, __LINE__,
+        cert.GetHash().ToString(), scIt->second.scInfo.topCommittedCertHash.ToString(), sidechainUndo.prevTopCommittedCertHash.ToString());
+    LogPrint("cert", "%s():%d - cert epoch %d, last epoch %d, undo epoch %d\n", __func__, __LINE__,
+        cert.epochNumber, scIt->second.scInfo.topCommittedCertReferencedEpoch, sidechainUndo.prevTopCommittedCertReferencedEpoch);
+
+    if (cert.GetHash() == scIt->second.scInfo.topCommittedCertHash)
```

Comment:
modify this into an assert (this method must be call only for block top quality certs)

Reply:
Done